File: serve_game.py
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler
from urllib.parse import urlsplit, parse_qs, unquote
from urllib.request import Request, urlopen
from urllib.error import HTTPError
from pathlib import Path
import json
import os
import re
import sys
import threading
import webbrowser

HOST = '0.0.0.0'
PORT = 8765
ROOT = Path(__file__).resolve().parent
GAME = 'gta_vice_city_destroy_and_kill_v31.html'
ALLOWED = [
    re.compile(r'(^|\.)arcgis\.com$', re.I),
    re.compile(r'(^|\.)arcgisonline\.com$', re.I),
    re.compile(r'(^|\.)esri\.com$', re.I),
    re.compile(r'(^|\.)iprpraha\.cz$', re.I),
    re.compile(r'(^|\.)praha\.eu$', re.I),
]

class Handler(SimpleHTTPRequestHandler):
    protocol_version = 'HTTP/1.1'

    def translate_path(self, path):
        clean = urlsplit(path).path
        if clean == '/':
            clean = '/' + GAME
        return str(ROOT / clean.lstrip('/'))

    def end_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET,HEAD,POST,OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type,Range,Accept,Authorization,X-Requested-With')
        self.send_header('Access-Control-Expose-Headers', 'Content-Length,Content-Range,Accept-Ranges,ETag,Last-Modified')
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Content-Length', '0')
        self.end_headers()

    def do_HEAD(self):
        if self.path.startswith('/proxy-health'):
            self._health(head=True)
        elif self.path.startswith('/arcgis-proxy'):
            self._proxy(head=True)
        else:
            super().do_HEAD()

    def do_GET(self):
        if self.path.startswith('/proxy-health'):
            self._health()
        elif self.path.startswith('/arcgis-proxy'):
            self._proxy()
        else:
            super().do_GET()

    def do_POST(self):
        if self.path.startswith('/arcgis-proxy'):
            self._proxy()
        else:
            self.send_error(405, 'Method not allowed')

    def _health(self, head=False):
        body = json.dumps({'ok': True, 'version': 30, 'proxy': True}).encode()
        self.send_response(200)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.send_header('Cache-Control', 'no-store')
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        if not head:
            self.wfile.write(body)

    def _target(self):
        parts = urlsplit(self.path)
        if parts.path.startswith('/arcgis-proxy/'):
            tail = unquote(parts.path[len('/arcgis-proxy/'):])
            if tail.startswith(('http://', 'https://')):
                return tail + (('?' + parts.query) if parts.query else '')
        raw = parts.query
        if raw.startswith('url='):
            return unquote(raw[4:])
        if raw.lower().startswith(('http%3a', 'https%3a')):
            return unquote(raw)
        if raw.startswith(('http://', 'https://')):
            return raw
        return parse_qs(raw).get('url', [''])[0]

    def _allowed(self, target):
        try:
            u = urlsplit(target)
            return u.scheme in ('http', 'https') and any(rx.search(u.hostname or '') for rx in ALLOWED)
        except Exception:
            return False

    def _proxy(self, head=False):
        target = self._target()
        if not self._allowed(target):
            self._json_error(400, 'Missing or disallowed proxy target', target)
            return
        length = int(self.headers.get('Content-Length') or 0)
        if length > 8 * 1024 * 1024:
            self._json_error(413, 'Proxy request body too large', target)
            return
        body = self.rfile.read(length) if length else None
        headers = {
            'User-Agent': self.headers.get('User-Agent', 'PragueGameLocalProxy/31'),
            'Accept': self.headers.get('Accept', '*/*'),
            'Accept-Encoding': self.headers.get('Accept-Encoding', 'identity'),
        }
        for name in ('Content-Type', 'Range', 'If-None-Match', 'If-Modified-Since'):
            if self.headers.get(name):
                headers[name] = self.headers[name]
        print('[proxy]', self.command, target, flush=True)
        try:
            req = Request(target, data=body, headers=headers, method='HEAD' if head else self.command)
            with urlopen(req, timeout=120) as upstream:
                payload = b'' if head else upstream.read()
                self.send_response(upstream.status)
                for name in ('Content-Type', 'Content-Encoding', 'Cache-Control', 'ETag', 'Last-Modified', 'Accept-Ranges', 'Content-Range'):
                    value = upstream.headers.get(name)
                    if value:
                        self.send_header(name, value)
                self.send_header('Content-Length', str(len(payload) if not head else int(upstream.headers.get('Content-Length') or 0)))
                self.end_headers()
                if not head:
                    self.wfile.write(payload)
        except HTTPError as err:
            payload = err.read()
            self.send_response(err.code)
            self.send_header('Content-Type', err.headers.get('Content-Type', 'application/octet-stream'))
            self.send_header('Content-Length', str(len(payload)))
            self.end_headers()
            if not head:
                self.wfile.write(payload)
        except Exception as err:
            print('[proxy error]', target, err, flush=True)
            self._json_error(502, str(err), target)

    def _json_error(self, status, message, target=''):
        payload = json.dumps({'error': {'message': message, 'target': target}}).encode()
        self.send_response(status)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.send_header('Content-Length', str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    def log_message(self, fmt, *args):
        if '/arcgis-proxy' not in self.path:
            super().log_message(fmt, *args)

os.chdir(ROOT)
server = ThreadingHTTPServer((HOST, PORT), Handler)
url = f'http://127.0.0.1:{PORT}/{GAME}'
print(f'DESTROY AND KILL v31 running at {url}', flush=True)
print('This launcher includes the Prague ArcGIS/IPR CORS proxy.', flush=True)
print('When Prague loads, proxy requests will appear in this window.', flush=True)
print('Keep this window open. Press Ctrl+C to stop.', flush=True)
threading.Timer(0.7, lambda: webbrowser.open(url)).start()
try:
    server.serve_forever()
except KeyboardInterrupt:
    pass
finally:
    server.server_close()

File: serve_game.js
const http = require('http');
const https = require('https');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

const host = '0.0.0.0';
const browserHost = '127.0.0.1';
const port = 8765;
const root = __dirname;
const game = 'gta_vice_city_destroy_and_kill_v31.html';
const mime = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.svg': 'image/svg+xml',
  '.bin': 'application/octet-stream'
};
const allowedHosts = [
  /(^|\.)arcgis\.com$/i,
  /(^|\.)arcgisonline\.com$/i,
  /(^|\.)esri\.com$/i,
  /(^|\.)iprpraha\.cz$/i,
  /(^|\.)praha\.eu$/i
];

function corsHeaders(extra = {}) {
  return {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET,HEAD,POST,OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type,Range,Accept,Authorization,X-Requested-With',
    'Access-Control-Expose-Headers': 'Content-Length,Content-Range,Accept-Ranges,ETag,Last-Modified',
    ...extra
  };
}

function parseProxyTarget(rawUrl) {
  const qIndex = rawUrl.indexOf('?');
  const pathname = decodeURIComponent((qIndex >= 0 ? rawUrl.slice(0, qIndex) : rawUrl));
  const rawQuery = qIndex >= 0 ? rawUrl.slice(qIndex + 1) : '';
  if (pathname.startsWith('/arcgis-proxy/')) {
    const tail = pathname.slice('/arcgis-proxy/'.length);
    if (/^https?:/i.test(tail)) return tail + (rawQuery ? '?' + rawQuery : '');
  }
  if (rawQuery.startsWith('url=')) return decodeURIComponent(rawQuery.slice(4));
  if (/^https?%3A/i.test(rawQuery)) return decodeURIComponent(rawQuery);
  if (/^https?:/i.test(rawQuery)) return rawQuery;
  try {
    const params = new URLSearchParams(rawQuery);
    const target = params.get('url');
    if (target) return target;
  } catch (_) {}
  return '';
}

function isAllowedTarget(target) {
  try {
    const url = new URL(target);
    return (url.protocol === 'https:' || url.protocol === 'http:') && allowedHosts.some(re => re.test(url.hostname));
  } catch (_) {
    return false;
  }
}

function collectBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    let size = 0;
    req.on('data', chunk => {
      size += chunk.length;
      if (size > 8 * 1024 * 1024) {
        reject(new Error('Proxy request body too large'));
        req.destroy();
        return;
      }
      chunks.push(chunk);
    });
    req.on('end', () => resolve(Buffer.concat(chunks)));
    req.on('error', reject);
  });
}

function proxyRequest(target, req, res, body, redirects = 0) {
  if (redirects > 5) {
    res.writeHead(502, corsHeaders({ 'Content-Type': 'text/plain; charset=utf-8' }));
    res.end('Too many upstream redirects');
    return;
  }
  const url = new URL(target);
  const transport = url.protocol === 'https:' ? https : http;
  const headers = {
    'User-Agent': req.headers['user-agent'] || 'PragueGameLocalProxy/31',
    'Accept': req.headers.accept || '*/*',
    'Accept-Encoding': req.headers['accept-encoding'] || 'identity'
  };
  for (const name of ['content-type', 'range', 'if-none-match', 'if-modified-since']) {
    if (req.headers[name]) headers[name] = req.headers[name];
  }
  if (body.length) headers['Content-Length'] = String(body.length);

  const upstream = transport.request(url, {
    method: req.method,
    headers,
    timeout: 120000
  }, upstreamRes => {
    const status = upstreamRes.statusCode || 502;
    if ([301, 302, 303, 307, 308].includes(status) && upstreamRes.headers.location) {
      upstreamRes.resume();
      const next = new URL(upstreamRes.headers.location, url).toString();
      proxyRequest(next, req, res, status === 303 ? Buffer.alloc(0) : body, redirects + 1);
      return;
    }
    const pass = {};
    for (const name of ['content-type', 'content-length', 'content-encoding', 'cache-control', 'etag', 'last-modified', 'accept-ranges', 'content-range']) {
      if (upstreamRes.headers[name] !== undefined) pass[name] = upstreamRes.headers[name];
    }
    if (!pass['cache-control']) pass['cache-control'] = 'public, max-age=3600';
    res.writeHead(status, corsHeaders(pass));
    if (req.method === 'HEAD') {
      upstreamRes.resume();
      res.end();
    } else {
      upstreamRes.pipe(res);
    }
  });
  upstream.on('timeout', () => upstream.destroy(new Error('Upstream timeout')));
  upstream.on('error', err => {
    console.error('[proxy error]', target, err.message);
    if (!res.headersSent) res.writeHead(502, corsHeaders({ 'Content-Type': 'application/json; charset=utf-8' }));
    if (!res.writableEnded) res.end(JSON.stringify({ error: { message: err.message, target } }));
  });
  if (body.length && req.method !== 'GET' && req.method !== 'HEAD') upstream.write(body);
  upstream.end();
}

function serveStatic(req, res) {
  const requestPath = decodeURIComponent((req.url || '/').split('?')[0]);
  const relative = requestPath === '/' ? game : requestPath.replace(/^\/+/, '');
  const filePath = path.resolve(root, relative);
  const rootPath = path.resolve(root);
  if (!filePath.startsWith(rootPath + path.sep) && filePath !== path.resolve(root, game)) {
    res.writeHead(403, corsHeaders()); res.end('Forbidden'); return;
  }
  fs.stat(filePath, (statErr, stat) => {
    if (statErr || !stat.isFile()) { res.writeHead(404, corsHeaders()); res.end('Not found'); return; }
    res.writeHead(200, corsHeaders({
      'Content-Type': mime[path.extname(filePath).toLowerCase()] || 'application/octet-stream',
      'Content-Length': stat.size,
      'Cache-Control': 'no-store'
    }));
    if (req.method === 'HEAD') res.end();
    else fs.createReadStream(filePath).pipe(res);
  });
}

const server = http.createServer(async (req, res) => {
  try {
    if (req.method === 'OPTIONS') {
      res.writeHead(204, corsHeaders()); res.end(); return;
    }
    if ((req.url || '').startsWith('/proxy-health')) {
      res.writeHead(200, corsHeaders({ 'Content-Type': 'application/json; charset=utf-8', 'Cache-Control': 'no-store' }));
      res.end(JSON.stringify({ ok: true, version: 31, proxy: true }));
      return;
    }
    if ((req.url || '').startsWith('/arcgis-proxy')) {
      const target = parseProxyTarget(req.url || '');
      if (!isAllowedTarget(target)) {
        res.writeHead(400, corsHeaders({ 'Content-Type': 'application/json; charset=utf-8' }));
        res.end(JSON.stringify({ error: { message: 'Missing or disallowed proxy target', target } }));
        return;
      }
      const body = await collectBody(req);
      console.log('[proxy]', req.method, target);
      proxyRequest(target, req, res, body);
      return;
    }
    if (req.method !== 'GET' && req.method !== 'HEAD') {
      res.writeHead(405, corsHeaders()); res.end('Method not allowed'); return;
    }
    serveStatic(req, res);
  } catch (err) {
    console.error(err);
    if (!res.headersSent) res.writeHead(500, corsHeaders({ 'Content-Type': 'text/plain; charset=utf-8' }));
    if (!res.writableEnded) res.end(err.message || 'Server error');
  }
});

server.listen(port, host, () => {
  const url = `http://${browserHost}:${port}/${game}`;
  console.log(`DESTROY AND KILL v31 running at ${url}`);
  console.log('This launcher includes the Prague ArcGIS/IPR CORS proxy.');
  console.log('For a phone on the same Wi-Fi, open http://YOUR-PC-IP:'+String(port)+'/'+game);
  console.log('When Prague loads, proxy requests will appear in this window.');
  console.log('Keep this window open. Press Ctrl+C to stop.');
  const command = process.platform === 'win32'
    ? `start "" "${url}"`
    : process.platform === 'darwin'
      ? `open "${url}"`
      : `xdg-open "${url}"`;
  exec(command, () => {});
});
